File: packages/shared/src/algorithms/compoundChildEligibility.ts
/**
 * Compound child eligibility — may an EXISTING library task be linked as a
 * sub-task of a compound?
 *
 * The one guard shared by every path that links an existing task under a
 * compound: the Task Detail save (web `editCompoundStructure`, iOS
 * `applyTaskEditPatch`), which refuses with the message; the board wizard's
 * staged-edit apply, which skips an ineligible link silently; and the
 * sub-task picker, which disables ineligible rows with the message.
 *
 * Nested compounds ARE allowed (a child can be another compound — see
 * docs/TASK_SYSTEM.md); only a link that would close a loop is refused.
 * Hidden wizard drafts (`createdInWizard`) are excluded by the picker via
 * `computeBrowsableTasks`, not here.
 *
 * Swift twin: `CompoundChildEligibility.linkProblem` (identical strings,
 * identical check order).
 */
import { TaskType } from '../constants/enums';
import type { CompoundChild } from '../types/compoundChild';
import type { Task } from '../types/task';
import { isGoalLessCounter } from './browsableTasks';
import { findTransitiveParentCompounds } from './derivationPass';

/** The fields of a candidate task the guard reads. */
export type CompoundChildCandidate = Pick<
  Task,
  'id' | 'type' | 'isDeleted' | 'createdInWizard' | 'isCounter' | 'maxCount'
>;

/** The six user-facing refusal messages (byte-identical on iOS). */
export const COMPOUND_CHILD_LINK_MESSAGES = {
  self: 'A compound can’t contain itself.',
  duplicate: 'That task is already a sub-task here.',
  achievement: 'Achievements can’t be sub-tasks.',
  deleted: 'That task was deleted.',
  goalLessCounter: 'Counters without a goal can’t be sub-tasks.',
  loop: 'That would create a loop — it already contains this compound.',
} as const;

/**
 * Returns null when `candidate` may be linked under `parentId`, else the
 * user-facing reason. Checks run in this order: self, duplicate,
 * achievement, deleted, goal-less counter, loop — the first failing check
 * wins. The goal-less check mirrors the `isGoalLessCounter` write guard every
 * other compound-child write enforces (a hub counter with no goal can't
 * evaluate as a sub-task).
 *
 * @param parentId - The compound the candidate would be linked under.
 * @param candidate - The existing task being linked.
 * @param allLinks - Live links across ALL compounds (soft-deleted rows are
 *   ignored by the loop walk).
 * @param currentChildIds - The editor's current kept children (existing +
 *   already-picked).
 * @returns `null` if eligible, otherwise one of {@link COMPOUND_CHILD_LINK_MESSAGES}.
 */
export function compoundChildLinkProblem(
  parentId: string,
  candidate: CompoundChildCandidate,
  allLinks: readonly CompoundChild[],
  currentChildIds: ReadonlySet<string>,
): string | null {
  if (candidate.id === parentId) return COMPOUND_CHILD_LINK_MESSAGES.self;
  if (currentChildIds.has(candidate.id)) return COMPOUND_CHILD_LINK_MESSAGES.duplicate;
  if (candidate.type === TaskType.ACHIEVEMENT) return COMPOUND_CHILD_LINK_MESSAGES.achievement;
  if (candidate.isDeleted) return COMPOUND_CHILD_LINK_MESSAGES.deleted;
  if (isGoalLessCounter(candidate)) return COMPOUND_CHILD_LINK_MESSAGES.goalLessCounter;
  // A loop closes iff the candidate already (transitively) contains the parent.
  if (findTransitiveParentCompounds(parentId, [...allLinks]).has(candidate.id)) {
    return COMPOUND_CHILD_LINK_MESSAGES.loop;
  }
  return null;
}

/**
 * A counting task the compound editor can't keep as a sub-task: no positive
 * goal (`maxCount`) or a blank unit. `validatePatch` refuses such a counting
 * sub-task on save, so the picker hides it rather than offering a row that
 * can only fail later.
 *
 * @param task - The candidate task.
 * @returns `true` when the task is COUNTING and lacks a goal or unit.
 */
export function isIncompleteCountingChild(
  task: Pick<Task, 'type' | 'maxCount' | 'unit'>,
): boolean {
  if (task.type !== TaskType.COUNTING) return false;
  return !(typeof task.maxCount === 'number' && task.maxCount > 0) || (task.unit ?? '').trim() === '';
}

/**
 * The compound sub-task quick-add row's candidate list: the browsable library
 * (pass `computeBrowsableTasks` output — it already hides wizard drafts and
 * deleted rows) narrowed to tasks that can be linked under `parentId` right
 * now (`compoundChildLinkProblem === null`) and that would survive the save's
 * validation (not {@link isIncompleteCountingChild}), ordered by title.
 *
 * Swift twin: `CompoundChildEligibility.pickerCandidates`.
 *
 * @param parentId - The compound being edited.
 * @param browsable - The browsable library tasks.
 * @param allLinks - Live links across ALL compounds.
 * @param currentChildIds - The editor's current kept children.
 * @returns The eligible tasks, sorted by lower-cased title (plain code-unit
 *   order, locale-independent so both platforms agree), then id.
 */
export function compoundChildPickerCandidates<T extends CompoundChildCandidate & Pick<Task, 'title' | 'unit'>>(
  parentId: string,
  browsable: readonly T[],
  allLinks: readonly CompoundChild[],
  currentChildIds: ReadonlySet<string>,
): T[] {
  return browsable
    .filter(
      (t) =>
        !isIncompleteCountingChild(t) &&
        compoundChildLinkProblem(parentId, t, allLinks, currentChildIds) === null,
    )
    .sort((a, b) => {
      const ta = a.title.toLowerCase();
      const tb = b.title.toLowerCase();
      if (ta !== tb) return ta < tb ? -1 : 1;
      return a.id < b.id ? -1 : a.id > b.id ? 1 : 0;
    });
}

File: apps/ios/OYBC/Services/CompoundChildEligibility.swift
import Foundation

/// Compound child eligibility — may an EXISTING library task be linked as a
/// sub-task of a compound?
///
/// Swift twin of @oybc/shared's `compoundChildEligibility.ts`
/// (`compoundChildLinkProblem`): identical strings, identical check order.
/// Shared by the Task Detail save (`applyTaskEditPatch`, which refuses with
/// the message), the board wizard's staged-edit apply (which skips an
/// ineligible link silently) and the sub-task picker (which disables
/// ineligible rows with the message).
///
/// Nested compounds ARE allowed (a child can be another compound — see
/// docs/TASK_SYSTEM.md); only a link that would close a loop is refused.
/// Hidden wizard drafts (`createdInWizard`) are excluded by the picker via
/// the browsable-tasks filter, not here.
enum CompoundChildEligibility {

    /// The six user-facing refusal messages (byte-identical on web).
    enum Message {
        static let selfContainment = "A compound can’t contain itself."
        static let duplicate = "That task is already a sub-task here."
        static let achievement = "Achievements can’t be sub-tasks."
        static let deleted = "That task was deleted."
        static let goalLessCounter = "Counters without a goal can’t be sub-tasks."
        static let loop = "That would create a loop — it already contains this compound."
    }

    /// Returns nil when `candidate` may be linked under `parentId`, else the
    /// user-facing reason. Checks run in this order: self, duplicate,
    /// achievement, deleted, goal-less counter, loop — the first failing
    /// check wins. The goal-less check mirrors the
    /// `BrowsableTasks.isGoalLessCounter` write guard every other
    /// compound-child write enforces.
    ///
    /// - Parameters:
    ///   - parentId: The compound the candidate would be linked under.
    ///   - candidate: The existing task being linked.
    ///   - allLinks: Live links across ALL compounds (soft-deleted rows are
    ///     ignored by the loop walk).
    ///   - currentChildIds: The editor's current kept children (existing +
    ///     already-picked).
    /// - Returns: nil if eligible, otherwise one of `Message`.
    static func linkProblem(
        parentId: String,
        candidate: Task,
        allLinks: [CompoundChild],
        currentChildIds: Set<String>
    ) -> String? {
        if candidate.id == parentId { return Message.selfContainment }
        if currentChildIds.contains(candidate.id) { return Message.duplicate }
        if candidate.type == .achievement { return Message.achievement }
        if candidate.isDeleted { return Message.deleted }
        if BrowsableTasks.isGoalLessCounter(candidate) { return Message.goalLessCounter }
        // A loop closes iff the candidate already (transitively) contains the parent.
        if DerivationPass.findTransitiveParentCompounds(changedTaskId: parentId, children: allLinks)
            .contains(candidate.id) {
            return Message.loop
        }
        return nil
    }

    /// A counting task the compound editor can't keep as a sub-task: no
    /// positive goal (`maxCount`) or a blank unit. `TaskEditPatch.validate`
    /// refuses such a counting sub-task on save, so the picker hides it.
    /// Twin of web `isIncompleteCountingChild`.
    ///
    /// - Parameter task: The candidate task.
    /// - Returns: `true` when the task is COUNTING and lacks a goal or unit.
    static func isIncompleteCountingChild(_ task: Task) -> Bool {
        guard task.type == .counting else { return false }
        let hasGoal = (task.maxCount ?? 0) > 0
        let hasUnit = !(task.unit ?? "").trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
        return !hasGoal || !hasUnit
    }

    /// The "+ Existing task…" picker's candidate list: the browsable library
    /// narrowed to tasks `linkProblem` accepts under `parentId` and that would
    /// survive save validation (not `isIncompleteCountingChild`), ordered by
    /// lower-cased title (plain code-unit order, locale-independent so it
    /// matches web), then id. Twin of web `compoundChildPickerCandidates`.
    ///
    /// - Parameters:
    ///   - parentId: The compound being edited.
    ///   - browsable: The browsable library (`TaskLibraryViewModel.browsableTasks`).
    ///   - allLinks: Live links across ALL compounds.
    ///   - currentChildIds: The editor's current kept children.
    /// - Returns: The eligible tasks.
    static func pickerCandidates(
        parentId: String,
        browsable: [Task],
        allLinks: [CompoundChild],
        currentChildIds: Set<String>
    ) -> [Task] {
        browsable
            .filter {
                !isIncompleteCountingChild($0)
                    && linkProblem(parentId: parentId, candidate: $0, allLinks: allLinks, currentChildIds: currentChildIds) == nil
            }
            .sorted { a, b in
                let ta = Array(a.title.lowercased().utf16)
                let tb = Array(b.title.lowercased().utf16)
                if ta != tb { return ta.lexicographicallyPrecedes(tb) }
                return Array(a.id.utf16).lexicographicallyPrecedes(Array(b.id.utf16))
            }
    }

    /// Case-insensitive title search over picker candidates (a blank query
    /// keeps every row). Twin of web `searchCompoundChildCandidates`.
    ///
    /// - Parameters:
    ///   - tasks: The candidates.
    ///   - query: The search text.
    /// - Returns: The candidates whose title contains `query`.
    static func searchCandidates(_ tasks: [Task], query: String) -> [Task] {
        let q = query.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        guard !q.isEmpty else { return tasks }
        return tasks.filter { $0.title.lowercased().contains(q) }
    }
}
